File: lib/services/google_drive_provider.dart
import 'package:google_sign_in/google_sign_in.dart';
import 'package:googleapis/drive/v3.dart' as drive;
import 'package:http/http.dart' as http;
import '../models/cloud_file.dart';
import 'cloud_storage_provider.dart';

class GoogleDriveProvider implements CloudStorageProvider {
  static const _scopes = [drive.DriveApi.driveFileScope];

  final GoogleSignIn _googleSignIn = GoogleSignIn(
    scopes: _scopes,
  );

  GoogleSignInAccount? _currentUser;
  drive.DriveApi? _driveApi;

  @override
  String get providerName => 'Google Drive';

  @override
  bool get isAuthenticated => _currentUser != null && _driveApi != null;

  @override
  Future<bool> signIn() async {
    try {
      final account = await _googleSignIn.signIn();
      if (account == null) {
        return false;
      }

      _currentUser = account;
      final authHeaders = await account.authHeaders;
      final authenticateClient = _GoogleAuthClient(authHeaders);
      _driveApi = drive.DriveApi(authenticateClient);

      return true;
    } catch (e) {
      print('Google Drive sign-in error: $e');
      return false;
    }
  }

  @override
  Future<void> signOut() async {
    await _googleSignIn.signOut();
    _currentUser = null;
    _driveApi = null;
  }

  @override
  Future<List<CloudFile>> listFiles({String? folderId}) async {
    if (_driveApi == null) throw Exception('Not authenticated');

    final query = folderId == null
        ? "'root' in parents and trashed = false"
        : "'$folderId' in parents and trashed = false";

    final fileList = await _driveApi!.files.list(
      q: query,
      orderBy: 'folder,name',
      spaces: 'drive',
      $fields: 'files(id,name,mimeType,size,modifiedTime)',
    );

    return (fileList.files ?? []).map((file) {
      final isDir =
          file.mimeType == 'application/vnd.google-apps.folder';
      return CloudFile(
        id: file.id!,
        name: file.name!,
        path: file.name!, // We'd need to build full path if needed
        isDirectory: isDir,
        size: file.size != null ? int.parse(file.size!) : null,
        modifiedTime: file.modifiedTime,
        mimeType: file.mimeType,
      );
    }).toList();
  }

  @override
  Future<List<int>> downloadFile(String fileId) async {
    if (_driveApi == null) throw Exception('Not authenticated');

    final response = await _driveApi!.files.get(
      fileId,
      downloadOptions: drive.DownloadOptions.fullMedia,
    ) as drive.Media;

    final bytes = <int>[];
    await for (var chunk in response.stream) {
      bytes.addAll(chunk);
    }

    return bytes;
  }

  @override
  Future<CloudFile> uploadFile({
    required String name,
    required List<int> content,
    String? parentFolderId,
    String? mimeType,
  }) async {
    if (_driveApi == null) throw Exception('Not authenticated');

    final file = drive.File();
    file.name = name;
    file.mimeType = mimeType;
    if (parentFolderId != null) {
      file.parents = [parentFolderId];
    }

    final media = drive.Media(
      Stream.value(content),
      content.length,
    );

    final uploadedFile = await _driveApi!.files.create(
      file,
      uploadMedia: media,
    );

    return CloudFile(
      id: uploadedFile.id!,
      name: uploadedFile.name!,
      path: uploadedFile.name!,
      isDirectory: false,
      size: content.length,
      mimeType: mimeType,
    );
  }

  @override
  Future<CloudFile> createFolder({
    required String name,
    String? parentFolderId,
  }) async {
    if (_driveApi == null) throw Exception('Not authenticated');

    final folder = drive.File();
    folder.name = name;
    folder.mimeType = 'application/vnd.google-apps.folder';
    if (parentFolderId != null) {
      folder.parents = [parentFolderId];
    }

    final createdFolder = await _driveApi!.files.create(folder);

    return CloudFile(
      id: createdFolder.id!,
      name: createdFolder.name!,
      path: createdFolder.name!,
      isDirectory: true,
      mimeType: 'application/vnd.google-apps.folder',
    );
  }

  @override
  Future<void> delete(String fileId) async {
    if (_driveApi == null) throw Exception('Not authenticated');
    await _driveApi!.files.delete(fileId);
  }

  @override
  Future<List<CloudFile>> search(String query) async {
    if (_driveApi == null) throw Exception('Not authenticated');

    final searchQuery = "name contains '$query' and trashed = false";
    final fileList = await _driveApi!.files.list(
      q: searchQuery,
      orderBy: 'folder,name',
      spaces: 'drive',
      $fields: 'files(id,name,mimeType,size,modifiedTime)',
    );

    return (fileList.files ?? []).map((file) {
      final isDir =
          file.mimeType == 'application/vnd.google-apps.folder';
      return CloudFile(
        id: file.id!,
        name: file.name!,
        path: file.name!,
        isDirectory: isDir,
        size: file.size != null ? int.parse(file.size!) : null,
        modifiedTime: file.modifiedTime,
        mimeType: file.mimeType,
      );
    }).toList();
  }

  @override
  Future<CloudFile> getFile(String fileId) async {
    if (_driveApi == null) throw Exception('Not authenticated');

    final file = await _driveApi!.files.get(
      fileId,
      $fields: 'id,name,mimeType,size,modifiedTime',
    ) as drive.File;

    final isDir = file.mimeType == 'application/vnd.google-apps.folder';
    return CloudFile(
      id: file.id!,
      name: file.name!,
      path: file.name!,
      isDirectory: isDir,
      size: file.size != null ? int.parse(file.size!) : null,
      modifiedTime: file.modifiedTime,
      mimeType: file.mimeType,
    );
  }
}

/// HTTP client with authentication headers
class _GoogleAuthClient extends http.BaseClient {
  final Map<String, String> _headers;
  final http.Client _client = http.Client();

  _GoogleAuthClient(this._headers);

  @override
  Future<http.StreamedResponse> send(http.BaseRequest request) {
    request.headers.addAll(_headers);
    return _client.send(request);
  }
}

File: react-native-writr/src/services/CloudProviders/GoogleDriveProvider.ts
import {GoogleSignin} from '@react-native-google-signin/google-signin';
import axios from 'axios';
import {CloudProvider, CloudFile} from '../../types';
import {CloudFileFactory} from '../../models/CloudFile';
import RNFS from 'react-native-fs';

/**
 * Google Drive cloud storage provider
 * Uses Google Sign-In and Google Drive API v3
 */
export class GoogleDriveProvider implements CloudProvider {
  name = 'Google Drive';
  isSignedIn = false;
  private accessToken: string | null = null;
  private baseUrl = 'https://www.googleapis.com/drive/v3';

  constructor() {
    // Configure Google Sign In
    GoogleSignin.configure({
      scopes: ['https://www.googleapis.com/auth/drive.file'],
      webClientId: 'YOUR_WEB_CLIENT_ID', // Replace with actual Web Client ID
    });
  }

  /**
   * Sign in to Google Drive
   */
  async signIn(): Promise<boolean> {
    try {
      await GoogleSignin.hasPlayServices();
      const userInfo = await GoogleSignin.signIn();
      const tokens = await GoogleSignin.getTokens();

      this.accessToken = tokens.accessToken;
      this.isSignedIn = true;

      return true;
    } catch (error) {
      console.error('Google Sign-In error:', error);
      return false;
    }
  }

  /**
   * Sign out from Google Drive
   */
  async signOut(): Promise<void> {
    try {
      await GoogleSignin.revokeAccess();
      await GoogleSignin.signOut();
      this.accessToken = null;
      this.isSignedIn = false;
    } catch (error) {
      console.error('Google Sign-Out error:', error);
    }
  }

  /**
   * List files in a folder
   */
  async listFiles(folderId?: string): Promise<CloudFile[]> {
    if (!this.accessToken) {
      throw new Error('Not signed in');
    }

    try {
      const query = folderId
        ? `'${folderId}' in parents and trashed=false`
        : `'root' in parents and trashed=false`;

      const response = await axios.get(`${this.baseUrl}/files`, {
        headers: {
          Authorization: `Bearer ${this.accessToken}`,
        },
        params: {
          q: query,
          fields:
            'files(id, name, mimeType, size, modifiedTime, parents)',
          orderBy: 'folder,name',
        },
      });

      return response.data.files.map((file: any) =>
        CloudFileFactory.create(
          file.id,
          file.name,
          file.name,
          file.mimeType === 'application/vnd.google-apps.folder',
          file.size ? parseInt(file.size, 10) : undefined,
          file.modifiedTime,
          file.mimeType,
        ),
      );
    } catch (error) {
      console.error('Error listing files:', error);
      return [];
    }
  }

  /**
   * Download a file
   */
  async downloadFile(
    fileId: string,
    destinationPath: string,
  ): Promise<boolean> {
    if (!this.accessToken) {
      throw new Error('Not signed in');
    }

    try {
      const response = await axios.get(
        `${this.baseUrl}/files/${fileId}?alt=media`,
        {
          headers: {
            Authorization: `Bearer ${this.accessToken}`,
          },
          responseType: 'arraybuffer',
        },
      );

      await RNFS.writeFile(
        destinationPath,
        Buffer.from(response.data).toString('base64'),
        'base64',
      );

      return true;
    } catch (error) {
      console.error('Error downloading file:', error);
      return false;
    }
  }

  /**
   * Upload a file
   */
  async uploadFile(
    localPath: string,
    fileName: string,
    parentFolderId?: string,
  ): Promise<string | null> {
    if (!this.accessToken) {
      throw new Error('Not signed in');
    }

    try {
      // Read file content
      const fileContent = await RNFS.readFile(localPath, 'base64');

      // Create metadata
      const metadata = {
        name: fileName,
        parents: parentFolderId ? [parentFolderId] : ['root'],
      };

      // Upload file (simplified - in production use multipart upload)
      const response = await axios.post(
        `${this.baseUrl}/files`,
        {
          ...metadata,
          data: fileContent,
        },
        {
          headers: {
            Authorization: `Bearer ${this.accessToken}`,
            'Content-Type': 'application/json',
          },
        },
      );

      return response.data.id;
    } catch (error) {
      console.error('Error uploading file:', error);
      return null;
    }
  }

  /**
   * Create a folder
   */
  async createFolder(
    name: string,
    parentFolderId?: string,
  ): Promise<CloudFile | null> {
    if (!this.accessToken) {
      throw new Error('Not signed in');
    }

    try {
      const metadata = {
        name,
        mimeType: 'application/vnd.google-apps.folder',
        parents: parentFolderId ? [parentFolderId] : ['root'],
      };

      const response = await axios.post(`${this.baseUrl}/files`, metadata, {
        headers: {
          Authorization: `Bearer ${this.accessToken}`,
          'Content-Type': 'application/json',
        },
        params: {
          fields: 'id, name, mimeType',
        },
      });

      return CloudFileFactory.create(
        response.data.id,
        response.data.name,
        response.data.name,
        true,
        undefined,
        undefined,
        response.data.mimeType,
      );
    } catch (error) {
      console.error('Error creating folder:', error);
      return null;
    }
  }

  /**
   * Delete a file
   */
  async deleteFile(fileId: string): Promise<boolean> {
    if (!this.accessToken) {
      throw new Error('Not signed in');
    }

    try {
      await axios.delete(`${this.baseUrl}/files/${fileId}`, {
        headers: {
          Authorization: `Bearer ${this.accessToken}`,
        },
      });

      return true;
    } catch (error) {
      console.error('Error deleting file:', error);
      return false;
    }
  }

  /**
   * Search for files
   */
  async search(query: string): Promise<CloudFile[]> {
    if (!this.accessToken) {
      throw new Error('Not signed in');
    }

    try {
      const searchQuery = `name contains '${query}' and trashed=false`;

      const response = await axios.get(`${this.baseUrl}/files`, {
        headers: {
          Authorization: `Bearer ${this.accessToken}`,
        },
        params: {
          q: searchQuery,
          fields:
            'files(id, name, mimeType, size, modifiedTime)',
          orderBy: 'name',
        },
      });

      return response.data.files.map((file: any) =>
        CloudFileFactory.create(
          file.id,
          file.name,
          file.name,
          file.mimeType === 'application/vnd.google-apps.folder',
          file.size ? parseInt(file.size, 10) : undefined,
          file.modifiedTime,
          file.mimeType,
        ),
      );
    } catch (error) {
      console.error('Error searching files:', error);
      return [];
    }
  }
}
